File: 프로그래머스/3/12920. 선입 선출 스케줄링/선입 선출 스케줄링.py
def solution(n, cores):
    def check_cores(t):
        return sum(t // core for core in cores)

    if n < len(cores):
        return n

    n -= len(cores)
    left, right = 1, max(cores) * n
    target_time = right

    while left < right:
        mid = (left + right) // 2
        if check_cores(mid) >= n:
            right = mid
        else:
            left = mid + 1

    target_time = left
    remain_work = n - sum((target_time - 1) // core for core in cores)

    for i, core in enumerate(cores, 1):
        if target_time % core != 0:
            continue
        
        remain_work -= 1
        if remain_work == 0:
            return i

File: 프로그래머스/3/12920. 선입 선출 스케줄링/선입 선출 스케줄링.java
import java.util.*;

class Solution {

    public int solution(int n, int[] cores) {
        int m = cores.length;
        if (n <= m) return n;

        int remain = n - m;
        long left = 1;
        long right = (long) Arrays.stream(cores).min().getAsInt() * remain;

        while (left < right) {
            long mid = (left + right) / 2;
            long done = 0;
            for (int c : cores) done += mid / c;
            if (done >= remain) right = mid;
            else left = mid + 1;
        }

        long doneBefore = 0;
        for (int c : cores) doneBefore += (left - 1) / c;

        long need = remain - doneBefore;
        for (int i = 0; i < m; i++) {
            if (left % cores[i] == 0 && --need == 0) return i + 1;
        }

        return -1;
    }
}